mu-plugins/enforce-2fa.php

```diff
@@ -0,0 +1,438 @@
+	if ( ! is_null( $set ) ) {
+		$is_jetpack_authorize_rest_request = $set;
+	}
+
+	return $is_jetpack_authorize_rest_request;
+}
+
+/**
+ * Hooked to the `rest_request_before_callbacks` filter to keep track of
+ * whether the current request is a REST API request for
+ * /jetpack/v4/remote_authorize request that connects WordPress.com user
+ * to a local user.
+ *
+ * @param WP_HTTP_Response $response The response object.
+ * @param array            $handler  The request handler.
+ *
+ * @return WP_HTTP_Response - it's attached to a filter.
+ */
+function e2fa_is_jetpack_authorize_rest_request_hook( $response, $handler ) {
+
+	// If the request is for /jetpack/v4/remote_authorize, set the flag.
+	if ( isset( $handler['callback'] ) && 'Jetpack_Core_Json_Api_Endpoints::remote_authorize' === $handler['callback'] ) {
+		e2fa_is_jetpack_authorize_rest_request( true );
+	}
+	return $response;
+}
+
+// Hook to keep track of whether the current request is a REST API request for /jetpack/v4/remote_authorize request that connects a WordPress.com user to a local user.
+add_filter( 'rest_request_before_callbacks', 'e2fa_is_jetpack_authorize_rest_request_hook', 10, 2 );
+
+/**
+ * Check if 2FA is forced.
+ *
+ * @return bool
+ */
+function e2fa_is_two_factor_forced() {
+
+	// If we're not forcing 2FA, return early.
+	if ( ! e2fa_should_force_two_factor() ) {
+		return false;
+	}
+
+	return apply_filters( 'e2fa_is_two_factor_forced', false );
+}
+
+/**
+ * Force 2FA.
+ */
+function e2fa_enforce_two_factor_plugin() {
+
+	// Check if user is logged in.
+	if ( is_user_logged_in() ) {
+
+		$cap     = apply_filters( 'two_factor_enforcement_cap', 'manage_options' );
+		$limited = current_user_can( $cap );
+
+		// Calculate current_user_can outside map_meta_cap to avoid callback loop.
+		add_filter(
+			'e2fa_is_two_factor_forced',
+			function () use ( $limited ) {
+				return $limited;
+			},
+			9
+		);
+
+		// If the user is not using 2FA, force them to use it.
+		$is_user_using_two_factor = Two_Factor_Core::is_user_using_two_factor();
+
+		add_filter(
+			'is_user_using_two_factor',
+			function () use ( $is_user_using_two_factor ) {
+				return $is_user_using_two_factor;
+			}
+		);
+
+		// Show admin notice, if user is not using 2FA.
+		add_action( 'admin_notices', 'e2fa_two_factor_admin_notice' );
+
+		// Filter the user's capabilities to remove the ones that would allow them to bypass 2FA.
+		add_filter( 'map_meta_cap', 'e2fa_two_factor_filter_caps', 0, 4 );
+	}
+}
+
+// If we're not installing, enable the two-factor plugin on muplugins_loaded.
+if ( ! defined( 'WP_INSTALLING' ) ) {
+
+	// Enable the two-factor plugin after all mu-plugins have been loaded.
+	add_action( 'muplugins_loaded', 'e2fa_enable_two_factor_plugin' );
+}
+
+/**
+ * Enable 2FA Plugin.
+ *
+ * @return void
+ */
+function e2fa_enable_two_factor_plugin() {
+
+	// If the two-factor plugin is not enabled, return early.
+	$enable_two_factor = apply_filters( 'enable_two_factor', true );
```

Comment:
Is this needed? If someone is using this as a mu-plugin then it should do something actionable else no need to keep this mu-plugin anyway.

Also since, mu-plugin is loaded very early, someone need to hook on this very early which a bit overkill IMO.

Reply:
Can we connect regarding this, when you are available?

Reply:
Removed for now 👍